File: JavaFundamentals/ExamPractice/src/org/softuni/LegendaryFarming.java
package org.softuni;

// Java Fundamentals Retake - 26 October 2015

import java.util.Map;
import java.util.Scanner;
import java.util.TreeMap;

public class LegendaryFarming {

    public static void main(String[] args) {

        Scanner in = new Scanner(System.in);
        Map<String, Integer> items = new TreeMap<>();
        Map<String, Integer> junk = new TreeMap<>();

        items.put("shards", 0);
        items.put("fragments", 0);
        items.put("motes", 0);

        while (true) {
            String[] input = in.nextLine().trim().toLowerCase().split("\\s+");

            for (int i = 0; i < input.length; i += 2) {
                String item = input[i + 1];
                Integer quantity = Integer.parseInt(input[i]);

                switch (item) {
                    case "shards":
                    case "fragments":
                    case "motes":
                            items.put(item, items.get(item) + quantity);
                        break;
                    default:
                        if (!junk.containsKey(item)) {
                            junk.put(item, quantity);
                        } else {
                            junk.put(item, junk.get(item) + quantity);
                        }
                        break;
                }

                if (farmingDone(items)) {
                    printMaps(items, junk);
                    return;
                }
            }
        }
    }

    private static void printMaps(Map<String, Integer> items, Map<String, Integer> junk) {
        items.entrySet().stream()
                .sorted((e1, e2) -> e2.getValue().compareTo(e1.getValue()))
                .forEach(e -> System.out.printf("%s: %d\r\n", e.getKey(), e.getValue()));

        junk.entrySet().forEach(e -> System.out.printf("%s: %d\r\n", e.getKey(), e.getValue()));
    }

    private static boolean farmingDone(Map<String, Integer> items) {

        for (Map.Entry<String, Integer> item : items.entrySet()) {
            if (item.getValue() >= 250) {
                items.put(item.getKey(), item.getValue() - 250);
                switch (item.getKey()) {
                    case "shards":
                        System.out.println("Shadowmourne obtained!");
                        break;
                    case "fragments":
                        System.out.println("Valanyr obtained!");
                        break;
                    case "motes":
                        System.out.println("Dragonwrath obtained!");
                        break;
                }
                return true;
            }
        }
        return false;
    }
}

File: C#-WebDeveloper-3.0/Advanced-C#-May-2016/Exercises/SetsAndDictionaries/LegendaryFarming/LegendaryFarming.cs
namespace BasicDictionaryOperations
{
    using System;
    using System.Collections.Generic;
    using System.Linq;

    class LegendaryFarming
    {
        static void Main(string[] args)
        {
            var items = new SortedDictionary<string, int>();
            var junk = new SortedDictionary<string, int>();

            items.Add("shards", 0);
            items.Add("fragments", 0);
            items.Add("motes", 0);

            while (true)
            {
                string[] input = Console.ReadLine()?
                    .Trim()
                    .ToLower()
                    .Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);

                for (int i = 0; i < input?.Length; i += 2)
                {
                    string item = input[i + 1];
                    int quantity = int.Parse(input[i]);

                    switch (item)
                    {
                        case "shards":
                        case "fragments":
                        case "motes":
                            items[item] += quantity;
                            break;
                        default:
                            if (!junk.ContainsKey(item))
                            {
                                junk.Add(item, quantity);
                            }
                            else
                            {
                                junk[item] += quantity;
                            }
                            break;
                    }

                    if (IsFarmingDone(items))
                    {
                        PrintLeftovers(items, junk);
                        return;
                    }
                }
            }
        }

        private static void PrintLeftovers(SortedDictionary<string, int> items, SortedDictionary<string, int> junk)
        {
            foreach (var item in items.OrderByDescending(i => i.Value))
            {
                Console.WriteLine($"{item.Key}: {item.Value}");
            }

            foreach (var pieceOfJunk in junk)
            {
                Console.WriteLine($"{ pieceOfJunk.Key}: { pieceOfJunk.Value}");
            }
        }

        private static bool IsFarmingDone(SortedDictionary<string, int> items)
        {
            var successMessages = new Dictionary<string, string>
                                      {
                                          { "shards", "Shadowmourne obtained!" },
                                          { "fragments", "Valanyr obtained!" },
                                          { "motes", "Dragonwrath obtained!" }
                                      };

            foreach (var item in items.Where(item => item.Value >= 250))
            {
                items[item.Key] -= 250;
                Console.WriteLine(successMessages[item.Key]);

                return true;
            }

            return false;
        }
    }
}